### DISCUSSION TITLE
Add bandcamp as a Music Provider

### DISCUSSION THREAD

**Discussion by WindowsMe2** [music-assistant/support#2315] [Feature requests and ideas]
I would love to see bandcamp being available as a music provider. 

https://bandcamp.com/

**Comment by pljones** [comment]
Ideally multiple account support, too.

**Comment by Zeronozero** [comment]
Yes, please add Bandcamp to MA, I have a thousand album on BC ! thanks

**Comment by infeeeee** [comment]
Mopidy has a Bandcamp integration, it's very stable: https://github.com/impliedchaos/mopidy-bandcamp

It's MIT licensed, also written in Python. Developers can take a look how it works under the hood. 

Afaik there is no official Bandcamp API for playback.

**Comment by Zeronozero** [comment]
Any possibility to add Bandcamp ? Thanks in advance

**Comment by spudje** [comment]
The lack of Bandcamp integration is really the only reason why I haven't switch from LMS to Music Assistant yet. The LMS Bandcamp integration allows you to stream any Bandcamp music, not just the music you purchased.

**Reply by pljones** [reply]
Does it understand "just the music you purchased" - that is, can it access all my music, even that which isn't available without purchase?

Oh, and I was trying to have a poke around in the [LMS source](https://github.com/epoupon/lms) -- is that the right "LMS", as there seem to be alternative uses ([Lyrion](https://github.com/lms-community/slimserver), for example, was styled "LMS").

**Reply by spudje** [reply]
This is the LMS version I'm running: Logitech Media Server Version: 8.5.2 - 1716215514 (docker lmscommunity/logitechmediaserver)
Bandcamp plugin: Band's Campout (v1.9.4)

It has this functionality:
![image](https://github.com/user-attachments/assets/b5e6ffec-a20d-464a-b6ae-c292b8064a78)

My bandcamp submenu gives you:
![image](https://github.com/user-attachments/assets/ffe16838-7dc9-4720-ae5a-d9eb8a905440)

**Reply by pljones** [reply]
OK, so that's Lyrion then:
> Lyrion Music Server (aka. LMS, fka. Logitech Mediaserver, SlimServer, SqueezeCenter, SqueezeboxServer, SliMP3)

And the link to the plugin: https://www.herger.net/slim-plugins/detail.php@nr=1341&kategorie=slim-plugins

**Reply by spudje** [reply]
Yes. They just changed the name for newer versions. I hadn't upgraded. Now I have from LMS to Lyrion. And indeed that's the (very good) Bandcamp plugin. If something similar, but much simpler is just fine (playing any Bandcamp music, frommy own collection but also not bought) would come to MA it would be awesome and I'd switch in an instance!

**Reply by pljones** [reply]
By the way, did you get the integration with the latest Lyrion working?  I enabled and configured the plugin but couldn't find it anywhere as a provider.

At some point, if I get time, I'll see if I can read the source to see how it works and then see if I can learn how to write a MA provider plugin...

**Reply by spudje** [reply]
Yes, the upgrade was easy, just start a container from the Lyrion docker image with the exact same config of the LMS docker container and it pciked up all my settings. Including the bandcamp and all other plugins.

**Reply by spudje** [reply]
It should show up under "My Apps"
![image](https://github.com/user-attachments/assets/cfbbb1b2-188d-4617-aa49-12b08cf6a0ef)

**Reply by pljones** [reply]
OK, got it -- <s>it's a fault in "Material Skin": "Apps" doesn't show the same as "My Apps".  "Default" works.</s>

Weird, after a few refreshes, it's appeared.

**Reply by pljones** [reply]
So, translating
https://github.com/LMS-Community/slimserver/blob/public/9.1/Slim/Plugin/OnlineLibrary/Plugin.pm
into something like
https://github.com/music-assistant/server/blob/dev/music_assistant/providers/ytmusic/__init__.py
along with all the faffling support bits doesn't look like a lot of work, to be honest.

**Reply by spudje** [reply]
> So, translating https://github.com/LMS-Community/slimserver/blob/public/9.1/Slim/Plugin/OnlineLibrary/Plugin.pm into something like https://github.com/music-assistant/server/blob/dev/music_assistant/providers/ytmusic/__init__.py along with all the faffling support bits doesn't look like a lot of work, to be honest.

That would be so awesome!

**Reply by pljones** [reply]
(Ongoing other projects are keeping me fully occupied... I retire in a few years, though...)

**Reply by pljones** [reply]
W00tz for CoPilot...

**WARNING**
I have no idea how to test this yet.

https://github.com/pljones/mamp-bandcamp/tree/main

OK, so I got CoPilot to port the code to save my head doing the unpicking and re-stitching.  It took a few goes (see the session log if you're really interested).

What I don't know is how to volume mount the new provider into the docker app.  So anyone reading who can help with that, please reply.

**Reply by pljones** [reply]
I'll be rude and ping @OzGav, as I asked CoPilot to use your YTMusic provider as a starting point.

Is it possible to test with the dockerised version or do I need the "bare" install?

----

**EDIT:** Oh I hate that.  It was so obvious:
```
$ docker container exec music-assistant-server find / -name manifest.json
/app/venv/lib/python3.12/site-packages/music_assistant/providers/podcastfeed/manifest.json
/app/venv/lib/python3.12/site-packages/music_assistant/providers/squeezelite/manifest.json
/app/venv/lib/python3.12/site-packages/music_assistant/providers/ytmusic/manifest.json
...
```
So adding a
`-v .../providers/Bandcamp:/app/venv/lib/python3.12/site-packages/music_assistant/providers/bandcamp`
to the docker start up might just be all I need...

**Reply by OzGav** [reply]
You need to setup a proper dev environment and test this before submitting a PR https://github.com/music-assistant/server/blob/dev/DEVELOPMENT.md

**Reply by pljones** [reply]
OK, never mind, what a mess.  Getting CoPilot not to get lost has taken years off me in the space of two hours...

Basically, it needs the API ported properly but it keeps dropping half the work rather than completing it, so it looks like it'll need finishing by hand.  At least it got as far as being recognised by MusicAssistant (and failing to reach Bandcamp with a 404...).

(And `vi` has been a proper dev environment since the 80s...)

**Reply by pljones** [reply]
Well, none of the AI tools seems able to get to grips with the problem.  https://github.com/pljones/mamp-bandcamp/tree/main has what's been generated so far - and it's senseless rubbish.  One day, all coders will be replaced by AI, of course.

CoPilot couldn't even raise a complaint about its own deficiencies - it had to ask me to do it:
https://github.com/orgs/community/discussions/157684

**Reply by spudje** [reply]
Thanks a lot already for all effort you've put into this. Unfortunately my coding skills are very limited, so I cannot hlp.

**Comment by spudje** [comment]
Mmm, hopefully somebody else is up for it. I just got my first Voice PE and that can only act as Spotify Connect player with MA.

**Reply by OzGav** [reply]
This is not true. You should have added the Voice PE via the Home Assistant player provider. Having said that I am not sure what this has to do with bandcamp

**Reply by spudje** [reply]
Sorry, I was a bit too quick here. My reasoning was as follows.

Ideally I'd like Bandcamp as a music provider in MA. But that's not available (yet?).

Alternatively however I can still use the Bandcamp mobile app and cast to compatible (I assume Google Cast) players, similarly as you can cast Spotify music to Spotify Connect players. Now when using Voice PE in combination with MA, MA turns it into a Spotify Connect player. However, but I can't verify, since I'm not running MA, I suppose MA does not turn the Voice PE in a Google Cast compatible player. If it would it would make it a player option to cast to from the bandcamp app. I was hoping to replace my Chromecast audios with Voice PEs, supporting both Spotify Connect and Google (hence also Bandcamp) casting. (Basically this: https://github.com/orgs/music-assistant/discussions/2354)

So as long as there is no Bandcamp support, in any form, in MA, I'm not willing to migrate from Lyrion (LMS) to MA, although I'm very tempted since it looks quite cool. Outside of MA scope would be squeezelite functionality added to Voice PE without needing MA, then I can play both Bandcamp and Spotify on there thanks to Lyrion/LMS plugins.

**Comment by spudje** [comment]
Maybe we can try a different AI to have a go? Claude or Codium?

**Comment by Fred-DTV** [comment]
https://github.com/impliedchaos/mopidy-bandcamp

Maybe this could help, it is a Mopidy BC integration which works, even without API

**Comment by ALERTua** [comment]
I have created a pull request that introduces Bandcamp Music Provider to the Music Assistant ecosystem: https://github.com/music-assistant/server/pull/2871
Here's the PR for the documentation: https://github.com/music-assistant/beta.music-assistant.io/pull/64
You can test it using `d5369777_music_assistant_dev` addon with this configuration:
```yaml
server_repo: ALERTua/fork_music_assistant_server@bc
```

**Comment by OzGav** [comment]
This will be available from 2.8.0beta7

**Comment by spudje** [comment]
So awesome this is supported now. Hope to migrate from LMS/Lyrion later this year!

**Comment by pljones** [comment]
Being able to browse my collection by artist so easily is very nice!

It hangs when I go into my wishlist or someone I'm following's collection -- presumably these aren't indexed by default and it's trying to pull everything back.  Is there any way to get the API to pull back only a limited number at a time?

**Reply by OzGav** [reply]
@teancom

**Comment by teancom** [comment]
@pljones Hmm. Can you give me your Bandcamp username and/or follow me on Bandcamp (I'm teancom there too) and I'll try to replicate? It works for me just fine, but maybe my friends/followers on there aren't sufficiently large / have large enough collections to expose the bug.

**Reply by pljones** [reply]
Done: https://bandcamp.com/teancom/followers

**Reply by teancom** [reply]
Well, the good news is that I was totally able to browse your collection from within Music Assistant. The better news is the *second* time I did, it was instant. The bad news is the first load took ~10 minutes. 🫠 So consider this bug verified, I'll look into what I can do to make it not terrible.

**Reply by pljones** [reply]
I just tried browsing your collection - it's not related to number of items, unless it's above some low number.

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
